### ISSUE TITLE
RecoverySourceHandler error message has source and target doc counts switched

### ISSUE THREAD

**Issue by henakamaMSFT** [elastic/elasticsearch#21893]
In all versions of Elasticsearch where I've checked the source code, there's a bug in the error message for when a shard can't recover from the primary because the number of docs differ.

It lists numDocsTarget as the doc count for the source node, and numDocsSource as the doc count for the target node, but it should be the other way around. Since the only way I've found to fix these problems is to remove the shard that we've deemed to be wrong, this can be really dangerous and cause removal of the wrong shard.

Relevant line of RecoverySourceHandler.java:

> if (numDocsTarget != numDocsSource) {
>     throw new IllegalStateException("try to recover " + request.shardId() + " from primary shard with sync id but number of docs differ: " + **numDocsTarget** + " (" + **request.sourceNode().getName()** + ", primary) vs " + **numDocsSource** + "(" + **request.targetNode().getName()** + ")");
> }

**Comment by s1monw** [comment]
looks like a bug, thanks for raising this. do you wanna open a PR for this?

**Comment by s1monw** [comment]
can you tell how you ran into this exception and what version you are using?

**Comment by henakamaMSFT** [comment]
Sure, be happy to open a PR. We're currently using version 1.7, working on upgrading.

After a node restarted, one of the shards failed to initialize with the error message I listed above. I'm guessing that a write was dropped on the replica at some earlier point, even though the overall operation reported success (if this is a known issue that is fixed in later versions, I'd be very interested to hear it). To fix this, we removed the shard files from the node reporting the wrong doc count and let it recover from the primary.

When checking doc counts with /_cat/shards, I noticed that the doc count didn't match up with the error message. Checked on the source code to see where the number was coming from and found the bug.
